In [1]:
from database.GraphDB import KnowledgeGraphDB
from retrieval.retrieval_hub import RetrivalHub
from query_and_ranking.query import Query
from query_and_ranking.rerank import Reranker
from agentic import Agent
from tqdm import tqdm

2024-06-09 02:00:36.245799: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-06-09 02:00:36.245823: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-06-09 02:00:36.246432: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2024-06-09 02:00:36.828192: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT


In [ ]:
import os 
from dotenv import load_dotenv
load_dotenv()

neo4j_uri = os.getenv('NEO4J_URI')
neo4j_user = os.getenv('NEO4J_USER')
neo4j_password = os.getenv('NEO4J_PASSWORD')

In [2]:
kg = KnowledgeGraphDB(uri=neo4j_uri, user=neo4j_user, password=neo4j_password)
retrieval_hub = RetrivalHub(kg)
reranker = Reranker()
cv_query = Query(kg, retrieval_hub)

In [3]:
chat_agent = Agent(kg, retrieval_hub, reranker, cv_query)

Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.


Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

In [4]:
import json
questions = json.loads(open('data/benchmark/qa_cv.json').read())

In [5]:
answers = []
i = 0

for question in tqdm(questions):
    prompt = question['prompt']
    answer = dict()
    answer['prompt'] = prompt
    answer['answer'] = question['answer']
    try:
        res = chat_agent.get_cv_from_jd(prompt)
        ans = res['files']
        answer['predict'] = [a.replace('data/raw/data/clean4.0/', '') for a in ans]
    except:
        answer['predict'] = []
    answers.append(answer)
    with open('data/benchmark/qa_cv_kg.json', 'w') as f:
        f.write(json.dumps(answers))
    i += 1


  0%|          | 0/39 [00:00<?, ?it/s]

Extracting Job Description


You are not running the flash-attention implementation, expect numerical differences.
/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/langchain_core/vectorstores.py:330: UserWarning: Relevance scores must be between 0 and 1, got [(Document(page_content='Languages:\n- English\nMajors:\n- bachelor, computer science, national economics university hanoi, (GPA: 3.43)\nWork:\n- lts group (0.5 years)\n- viettel telecom (0.5 years)\nRoles:\n- software developer (0.5 years)\nSkills:\n- web development\n- api development\n- content management system\n- user interface design\n- artificial intelligence\n- database management\n- testing\n- deployment\n- english teaching\nProgramming languages:\n- java (0.5 years)\n- javascript (0.5 years)\n- typescript (0.5 years)\n- php (0.5 years)\n- spring boot (0.5 years)\n- react (0.5 years)\n- node (0.5 years)\n- graphql (0.5 years)\nCities:\n- ha noi\n', metadata={'id': 1593}), -208.47142856414308), (Document(page_content='Languages:\n- English\nM

End LLM
New Query:  {'cityList': ['ha noi'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['marketing manager', 'marketing analyst', 'marketing assistant', 'marketing officer']}
Extended Num Application:  15
Not enough application
New Query:  {'cityList': ['ha noi'], 'roleList': ['marketing manager', 'marketing assistant', 'marketing analyst', 'marketing', 'marketing officer']}
[786, 1371, 1513, 1648, 1802, 2109, 424]
Less -> Sematic Search
Using Sematic Search
Num CV ok:  12
Rerank
[1240, 1371, 1488, 1513, 1593, 1648, 1802, 2004, 2109, 222, 424, 786]
Time: 10.246462345123291
Extracting Job Description


/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/langchain_core/vectorstores.py:330: UserWarning: Relevance scores must be between 0 and 1, got [(Document(page_content='Languages:\n- English\nMajors:\n- undergraduate, information technology, ho chi minh city open university, (GPA: 3.6)\nWork:\n- m q -h o use (1.5 years)\n- healthy heart project (0.5 years)\nRoles:\n- mobile app developer (0.5 years)\n- web developer (0 years)\n- software developer (0 years)\nSkills:\n- web development\n- mobile app development\n- customer service\n- marketing\n- database design\n- api development\nProgramming languages:\n- java (0 years)\n- javascript (0 years)\n- python (0 years)\n- spring mvc (0 years)\n- react (0 years)\n- react native (0 years)\n- mysql (0 years)\n- html (0 years)\n- css (0 years)\nCities:\n- ho chi minh\n', metadata={'id': 1240}), -193.27409230280102), (Document(page_content='Languages:\n- English\nMajors:\n- bachelor, information security, ho chi minh city university of

End LLM
New Query:  {'cityList': ['ho chi minh'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['marketing manager', 'marketing analyst', 'marketing assistant', 'marketing officer']}
Extended Num Application:  7
Not enough application
New Query:  {'cityList': ['ho chi minh'], 'roleList': ['marketing manager', 'marketing assistant', 'marketing analyst', 'marketing', 'marketing officer']}
[92, 468]
Less -> Sematic Search
Using Sematic Search
Num CV ok:  5
Rerank
[1240, 1715, 1841, 468, 92]
Time: 7.893522024154663
Extracting Job Description
End LLM
New Query:  {'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['marketing executive', 'marketing manager', 'marketing analyst', 'marketing assistant'], 'role_exp': 0.0}
Extended Num Application:  10
Num CV ok:  8
Rerank
Reranking


  8%|▊         | 3/39 [00:26<05:04,  8.45s/it]

[972, 8, 1513, 468, 1749, 1648, 786]
Time: 7.957111835479736
Extracting Job Description


 10%|█         | 4/39 [00:35<05:10,  8.87s/it]

End LLM
New Query:  {'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['odoo developer'], 'suitableList': ['java developer', ' web developer', 'android developer', 'developer', 'python developer', 'embedded developer', 'it developer', 'software developer', 'software engineer']}
Extended Num Application:  3
Not enough application
New Query:  {'suitableList': ['java developer', ' web developer', 'android developer', 'developer', 'python developer', 'software engineer', 'embedded developer', 'it developer', 'software developer', 'back-end developer']}
[651, 262, 250, 642, 50, 850, 658, 1924, 708, 1579, 2073, 1917, 373, 406, 573, 1703, 751, 1500, 37, 705, 666, 602, 296, 200, 104, 1953, 1896, 1774, 1616, 1021, 776, 481, 409, 358, 345, 310, 279, 270, 60, 2027, 2023, 2006, 1941, 1891, 1841, 1493, 1488, 1428, 1410, 1308, 1127, 1104, 648, 92, 70, 64, 1974, 1593, 1462, 1240, 587, 83, 1715, 557, 550, 2103, 42, 2093, 2004, 222, 1313, 324, 191, 287, 1

/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/langchain_core/vectorstores.py:330: UserWarning: Relevance scores must be between 0 and 1, got [(Document(page_content='Languages:\n- English\nMajors:\n- bachelor, marketing management, national economics university, (GPA: 3.75)\nWork:\n- canifa fashion store (0.5 years)\n- lam pham co., ltd. (0.5 years)\nRoles:\n- marketing (1.5 years)\n- social media manager (0.5 years)\n- content writer (0.5 years)\n- sales associate (0.5 years)\n- salesperson (0.5 years)\nSkills:\n- research\n- information synthesis\n- team leadership\n- writing\n- social media management\n- content creation\n- marketing strategy\n- google forms\nCities:\n- ha noi\nCertifications:\n- International English Language Testing System (IELTS)\n- International Computer Driving Licence (ICDL)\nAwards:\n- International English Language Testing System (IELTS)\n- International Computer Driving Licence (ICDL)\n- Excellent Academic Achievement Award\n', metadata={'id': 

End LLM
New Query:  {'institutionList': ['national economics university (neu)'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['marketing manager', 'marketing analyst', 'marketing assistant', 'marketing officer']}
Extended Num Application:  7
Not enough application
New Query:  {'institutionList': ['national economics university (neu)', 'university of economics'], 'roleList': ['marketing manager', 'marketing assistant', 'marketing analyst', 'marketing', 'marketing officer']}
[424, 2109]
Less -> Sematic Search
Using Sematic Search
Num CV ok:  5
Rerank
[1274, 1513, 2109, 395, 424]
Time: 9.94589900970459
Extracting Job Description
End LLM
New Query:  {'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['data analyst', 'research scientist', 'data science / artificial intelligence', 'data scientist'], 'role_exp': 0.5}
Extended Num Application:  14
Reduce CV  19  ->  14  CV
New Query:  {'eduList'

 15%|█▌        | 6/39 [00:55<05:08,  9.35s/it]

[638, 606, 383, 280, 270, 2141, 2013, 1291, 1428, 1456, 1847]
Time: 9.516389846801758
Extracting Job Description


 18%|█▊        | 7/39 [01:05<05:08,  9.63s/it]

End LLM
New Query:  {'institutionList': ['hanoi university of industry', 'fpt polytechnic hà nội'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor']}
Extended Num Application:  7
Num CV ok:  6
Rerank
Reranking
[1802, 651, 1896, 2004, 2006]
Time: 10.212158441543579
Extracting Job Description


 21%|██        | 8/39 [01:12<04:37,  8.95s/it]

End LLM
New Query:  {'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['accountant', 'auditor']}
Extended Num Application:  4
Reduce CV  7  ->  4  CV
New Query:  {'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['auditor']}
Risk:  True
Num CV ok:  3
Rerank
[1817, 305, 738]
Time: 7.501044750213623
Extracting Job Description


 23%|██▎       | 9/39 [01:20<04:12,  8.42s/it]

End LLM
New Query:  {'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['account management', 'finance', 'accountant', 'auditor']}
Extended Num Application:  3
Reduce CV  8  ->  3  CV
New Query:  {'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['accountant', 'auditor']}
Still more -> Reduce CV
Risk:  True
Num CV ok:  3
Rerank
Reranking
[1857, 138]
Time: 7.244226932525635
Extracting Job Description


 26%|██▌       | 10/39 [01:27<03:56,  8.17s/it]

End LLM
New Query:  {'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['hr', 'hr generalist', 'hr executive', 'hr administrator'], 'role_exp': 0.0, 'suitableList': ['marketing assistant', 'hr administrator', 'hr manager', 'hr generalist', 'hr executive', 'hr', 'recruitment consultant', 'communications assistant', 'research assistant', 'office assistant']}
Extended Num Application:  1
Num CV ok:  1
Rerank
[1066]
Time: 7.601626634597778
Extracting Job Description


--- Logging error ---
Traceback (most recent call last):
  File "/home/quanghung20gg/anaconda3/lib/python3.11/logging/__init__.py", line 1110, in emit
    msg = self.format(record)
          ^^^^^^^^^^^^^^^^^^^
  File "/home/quanghung20gg/anaconda3/lib/python3.11/logging/__init__.py", line 953, in format
    return fmt.format(record)
           ^^^^^^^^^^^^^^^^^^
  File "/home/quanghung20gg/anaconda3/lib/python3.11/logging/__init__.py", line 687, in format
    record.message = record.getMessage()
                     ^^^^^^^^^^^^^^^^^^^
  File "/home/quanghung20gg/anaconda3/lib/python3.11/logging/__init__.py", line 377, in getMessage
    msg = msg % self.args
          ~~~~^~~~~~~~~~~
TypeError: not all arguments converted during string formatting
Call stack:
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/ipykernel_launcher.py", line 17, in <module>
    app.launc

End LLM
New Query:  {'institutionList': ['national economics university (neu)'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['data analyst', 'data analysis', 'analyst', 'data science / artificial intelligence', 'data scientist', 'research scientist', 'financial market data analyst'], 'suitableList': ['data analyst', 'data analysis', 'analyst', 'data science / artificial intelligence', 'machine learning engineer', 'data scientist', 'research analyst', 'research scientist', 'financial market data analyst']}
Extended Num Application:  12
Not enough application
New Query:  {'institutionList': ['national economics university (neu)', 'university of economics'], 'suitableList': ['data analyst', 'data analysis', 'analyst', 'business analyst', 'data science / artificial intelligence', 'scientist', 'machine learning engineer', 'data scientist', 'research analyst', 'sales analyst', 'research scientist', 'financial market data analyst']}
[1214,

 28%|██▊       | 11/39 [01:40<04:25,  9.50s/it]

[1214, 124, 1291, 1690, 1941, 2013, 732, 280, 561, 495, 463, 606]
Time: 12.516016244888306
Extracting Job Description


/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/langchain_core/vectorstores.py:330: UserWarning: Relevance scores must be between 0 and 1, got [(Document(page_content='Languages:\n- English\nMajors:\n- undergraduate, software development, fpt polytechnic college, (GPA: 3.44)\nRoles:\n- java developer (0 years)\n- full stack developer (0 years)\nSkills:\n- web development\n- back\n- end development\n- front\n- database design\n- api integration\n- business analysis\nProgramming languages:\n- java (0 years)\n- spring boot (0 years)\n- jpa (0 years)\n- react (0 years)\n- mysql (0 years)\n- html (0 years)\n- css (0 years)\n- javascript (0 years)\nCities:\n- ha noi\nCertifications:\n- Basic English Certificate\nAwards:\n- Basic English Certificate\n', metadata={'id': 648}), -147.22910409486573), (Document(page_content='Languages:\n- English\nMajors:\n- undergraduate, computer programming, fpt polytechnic hà nội, (GPA: 3.6)\nWork:\n- công ty cổ phần thương mại và phát triển công n

End LLM
New Query:  {'cityList': ['ha noi'], 'institutionList': ['fpt university', 'fpt polytechnic college', 'fpt university hcm'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor']}
Extended Num Application:  14
Not enough application
New Query:  {'cityList': ['ha noi'], 'institutionList': ['fpt university', 'fpt polytechnic college', 'fpt university hcm']}
[862, 776, 648, 1703, 1749, 64, 279, 358]
Less -> Sematic Search
Using Sematic Search
Num CV ok:  11
Rerank
[1593, 1703, 1749, 2006, 2051, 279, 358, 64, 648, 776, 862]
Time: 8.770936250686646
Extracting Job Description


 33%|███▎      | 13/39 [01:56<03:46,  8.71s/it]

End LLM
New Query:  {'languageList': ['English'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['content marketing manager', 'marketing manager', 'digital marketing manager', 'marketing officer']}
Extended Num Application:  4
Reduce CV  5  ->  4  CV
New Query:  {'languageList': ['English'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['marketing manager', 'digital marketing manager', 'marketing officer']}
Risk:  True
Num CV ok:  4
Rerank
Reranking
[972, 1139, 468]
Time: 7.400040149688721
Extracting Job Description
End LLM
New Query:  {'languageList': ['English'], 'institutionList': ['national economics university (neu)'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['data analyst', 'machine learning engineer', 'data scientist', 'dataops engineer'], 'suitableList': ['data analyst', 'data analysis', 'machine learning engineer', 'dat

 36%|███▌      | 14/39 [02:07<03:54,  9.39s/it]

[1690, 561, 606]
Time: 10.960225582122803
Extracting Job Description


 38%|███▊      | 15/39 [02:15<03:38,  9.11s/it]

End LLM
New Query:  {'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['full stack developer', 'front end developer', 'back-end developer', 'developer'], 'role_exp': 2.5, 'programming_exp': 2.5, 'suitableList': ['full stack developer', 'front end developer', 'developer', 'it developer', 'net developer', 'back-end developer']}
Extended Num Application:  3
Reduce CV  5  ->  3  CV
New Query:  {'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['front end developer', 'back-end developer'], 'role_exp': 2.5, 'programming_exp': 2.5, 'suitableList': ['full stack developer', 'front end developer', 'back-end developer', 'developer']}
Risk:  True
Num CV ok:  3
Rerank
Reranking
[1579, 2083]
Time: 8.459286451339722
Extracting Job Description


/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/langchain_core/vectorstores.py:330: UserWarning: Relevance scores must be between 0 and 1, got [(Document(page_content='Languages:\n- English\nMajors:\n- bachelor, information security, ho chi minh city university of technology and food industry, (GPA: 2.9)\nWork:\n- công ty c ổ  ph ầ n công ngh ệ tam khoadeveloper .net (1 years)\nRoles:\n- tester (0 years)\n- researcher (0.5 years)\n- software developer (1 years)\nSkills:\n- software development\n- machine learning\n- sql\n- technical support\nProgramming languages:\n- .net (1 years)\n- python (0.5 years)\n- c# (1 years)\n- sql (1 years)\nCities:\n- ho chi minh\n', metadata={'id': 1841}), -161.0662610084791), (Document(page_content='Languages:\n- English\n- Vietnamese\nMajors:\n- undergraduate, computer networking and communication, ho chi minh city university of technology, (GPA: 3.6)\nWork:\n- hung phat usa (0.5 years)\n- private supermarket (0.5 years)\n- mwg (0.5 years)\n-

End LLM
New Query:  {'cityList': ['ho chi minh'], 'institutionList': ['ho chi minh city university of industry and trade', 'ho chi minh city industry and trade college'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor']}
Extended Num Application:  6
Not enough application
New Query:  {'cityList': ['ho chi minh'], 'institutionList': ['ho chi minh city university of industry and trade', 'ho chi minh city industry and trade college']}
[178, 191]
Less -> Sematic Search
Using Sematic Search
Num CV ok:  4
Rerank
[1530, 178, 1841, 191]
Time: 8.248933792114258
Extracting Job Description


 44%|████▎     | 17/39 [02:30<03:01,  8.23s/it]

End LLM
New Query:  {'institutionList': ['ho chi minh city open university'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor']}
Extended Num Application:  7
Num CV ok:  6
Rerank
Reranking
[548, 1193, 1669, 1240, 1604]
Time: 6.796512603759766
Extracting Job Description
End LLM
New Query:  {'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['data analyst', 'data analysis', 'analyst', 'business analyst', 'finance analyst', 'sales analyst', 'financial market data analyst'], 'role_exp': 0.5, 'suitableList': ['data analyst', 'data analysis', 'analyst', 'business analyst', 'marketing analyst', 'data scientist', 'finance analyst', 'research analyst', 'sales analyst', 'financial market data analyst']}
Extended Num Application:  7
Reduce CV  18  ->  7  CV
New Query:  {'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['data analyst', 'business analyst', 'financial mark

 46%|████▌     | 18/39 [02:38<02:50,  8.12s/it]

[1473, 335, 135, 1847, 1428]
Time: 7.859335422515869
Extracting Job Description


 49%|████▊     | 19/39 [02:46<02:38,  7.94s/it]

End LLM
New Query:  {'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['embedded developer', 'developer', 'embedded systems engineer', 'it developer']}
Extended Num Application:  3
Reduce CV  18  ->  3  CV
New Query:  {'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['embedded developer']}
Risk:  True
Not enough application
Reduce Risk -> Get more CV
New Query:  {'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['embedded developer', 'developer', 'embedded systems engineer']}
[850, 642, 587, 409, 345, 92, 23, 1953, 1493, 1500, 1669, 705, 2073, 279, 83, 1715, 1703]
More -> Reduce CV
Num CV ok:  3
Rerank
Reranking
[642, 1953]
Time: 7.5302934646606445
Extracting Job Description


 51%|█████▏    | 20/39 [02:54<02:32,  8.04s/it]

End LLM
New Query:  {'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['it support'], 'suitableList': ['technical team leader', 'media specialist', 'communication specialist', 'receptionist', 'customer service manager', 'e-commerce specialist', 'it support', 'change management specialist', 'customer service staff', 'research assistance', 'office assistant']}
Extended Num Application:  3
Num CV ok:  2
Rerank
[1530, 178]
Time: 8.278629779815674
Extracting Job Description


 54%|█████▍    | 21/39 [03:02<02:23,  7.98s/it]

End LLM
New Query:  {'cityList': ['ha noi'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': [' web developer', 'developer', 'mobile app developer', 'android developer']}
Extended Num Application:  4
Reduce CV  16  ->  4  CV
New Query:  {'cityList': ['ha noi'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': [' web developer', 'developer', 'mobile app developer', 'android developer']}
Still more -> Reduce CV
Risk:  True
Num CV ok:  4
Rerank
Reranking
[222, 279, 1896]
Time: 7.834760904312134
Extracting Job Description


/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/langchain_core/vectorstores.py:330: UserWarning: Relevance scores must be between 0 and 1, got [(Document(page_content='Languages:\n- English\nMajors:\n- bachelor, software engineer, hanoi university of industry, (GPA: 3.5)\nRoles:\n- android developer (0 years)\nSkills:\n- user interface design\n- api design\n- firebase\n- android development\n- git\n- postman\n- mvvm architecture\n- clean architecture\nProgramming languages:\n- kotlin (0 years)\n- dart (0 years)\nCities:\n- ha noi\nAwards:\n- Second Prize in Japanese dubbing contest\n- Consolation Prize in School-level Olympic Programming Contest\n- Third Prize in School-level Olympic Programming Contest\n', metadata={'id': 651}), -174.7689144773811), (Document(page_content='Languages:\n- English\nMajors:\n- bachelor, computer science, national economics university hanoi, (GPA: 3.43)\nWork:\n- lts group (0.5 years)\n- viettel telecom (0.5 years)\nRoles:\n- software developer 

End LLM
New Query:  {'institutionList': ['fpt polytechnic hà nội', 'university of information technology'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor']}
Extended Num Application:  8
Not enough application
New Query:  {'institutionList': ['fpt polytechnic hà nội', 'university of information technology']}
[363, 2006]
Less -> Sematic Search
Using Sematic Search
Num CV ok:  6
Rerank
[1593, 19, 2006, 222, 363, 651]
Time: 8.005748510360718
Extracting Job Description
End LLM
New Query:  {'cityList': ['ho chi minh'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['full stack developer', 'front end developer', 'back-end developer', 'developer'], 'role_exp': 0.0}
Extended Num Application:  7
Reduce CV  20  ->  7  CV
New Query:  {'cityList': ['ho chi minh'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['full stack developer', 'front end developer', 'back-

 59%|█████▉    | 23/39 [03:21<02:22,  8.88s/it]

[751, 708, 1715, 439, 1021]
Time: 10.955459594726562
Extracting Job Description


/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/langchain_core/vectorstores.py:330: UserWarning: Relevance scores must be between 0 and 1, got [(Document(page_content='Languages:\n- English\nMajors:\n- bachelor, software engineer, fpt university, (GPA: 3.6)\nRoles:\n- developer (0 years)\nSkills:\n- software development\n- teamwork\n- problem solving\nProgramming languages:\n- c# (0 years)\n- java (0 years)\n- asp.net mvc (0 years)\n- asp.net core api (0 years)\n- sql server (0 years)\n- mongodb (0 years)\n- angular (0 years)\n- bootstrap (0 years)\nCities:\n- cam pha\n', metadata={'id': 1953}), -275.0074835913298), (Document(page_content='Languages:\n- English\nMajors:\n- bachelor, information technology, tien giang university, (GPA: 3.6)\nWork:\n- sharework corp (0.5 years)\nRoles:\n- software developer (0 years)\n-  (0.5 years)\nSkills:\n- user interface design\n- mobile app development\n- web development\n- database management\n- api integration\n- figma\n- linux\n- dock

End LLM
New Query:  {'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['front end developer', 'developer', 'full stack developer', 'back-end developer'], 'role_exp': 5.0}
Extended Num Application:  6
Not enough application
New Query:  {'roleList': ['full stack developer', 'front end developer', 'developer', 'it developer', 'back-end developer'], 'role_exp': 4.5}
[502, 708, 23]
Less -> Sematic Search
Using Sematic Search
Num CV ok:  4
Rerank
[1953, 23, 502, 708]
Time: 8.266287803649902
Extracting Job Description


/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/langchain_core/vectorstores.py:330: UserWarning: Relevance scores must be between 0 and 1, got [(Document(page_content='Languages:\n- English\nWork:\n- dong tien service corporation (5 years)\n- phu nhuan jewelry (1 years)\n- fe credit (2 years)\n- fe credit (2.5 years)\n- sieuviet (1.5 years)\n- sieu viet (1.5 years)\nRoles:\n- back end developer (5 years)\n- java developer (1 years)\n- technical team leader (1 years)\n- solution architect (4.5 years)\n- dataops engineer (1.5 years)\n- devops engineer (5.5 years)\nSkills:\n- devops\n- cloud architecture\n- ci\n- cd\n- infrastructure as code\n- security\n- monitoring\n- automation\n- containerization\nProgramming languages:\n- java (5.5 years)\n- python (5.5 years)\n- groovy (5.5 years)\n- typescript (5.5 years)\n- sql (5.5 years)\n- spring boot (5.5 years)\n- linux bashshell (5.5 years)\n', metadata={'id': 502}), -247.43179554100547), (Document(page_content='Languages:\n- Engl

End LLM
New Query:  {'languageList': ['English'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['devops', 'developer', 'dataops engineer'], 'role_exp': 1.0, 'suitableList': ['developer', 'devops', 'back-end developer', 'it developer', 'dataops engineer', 'software engineer']}
Extended Num Application:  4
Reduce CV  9  ->  4  CV
New Query:  {'languageList': ['English'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['devops'], 'role_exp': 1.0, 'suitableList': ['developer', 'devops', 'it developer', 'dataops engineer', 'software engineer']}
Risk:  True
Not enough application
Reduce Risk -> Get more CV
New Query:  {'languageList': ['English'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['devops', 'dataops engineer'], 'role_exp': 1.0, 'suitableList': ['developer', 'devops', 'back-end developer', 'it developer', 'dataops engineer', 'sof

 67%|██████▋   | 26/39 [03:46<01:53,  8.71s/it]

End LLM
New Query:  {'languageList': ['English'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': [' web developer', 'php developer', 'developer', 'java developer'], 'role_exp': 2.5, 'suitableList': ['java developer', 'developer', 'python developer', 'it developer', 'software developer', 'software engineer']}
Extended Num Application:  4
Num CV ok:  3
Rerank
[1579, 2023, 587]
Time: 9.182329893112183
Extracting Job Description


 69%|██████▉   | 27/39 [03:55<01:44,  8.74s/it]

End LLM
New Query:  {'cityList': ['ha noi'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['marketing manager', 'content writer', 'marketing assistant', 'marketing analyst', 'financial planner', 'production planner', 'content creator', 'content marketing', 'designer', 'strategic planning', 'marketing officer'], 'role_exp': 1.5, 'skillList': ['social media marketing', 'marketing strategy', 'content creation', 'content writing', 'marketing', 'content marketing', 'digital marketing', 'creative concept development', 'financial planning', 'email marketing', 'value creation', 'decision making', 'content management', 'planning', 'strategic planning', 'production planning', 'building'], 'suitableList': ['insurance advisor', 'marketing executive', 'event coordinator', 'marketing manager', 'content writer', 'marketing assistant', 'marketing analyst', 'content marketing manager', 'financial planner', 'production planner', 'content creator', 'con

 72%|███████▏  | 28/39 [04:04<01:37,  8.90s/it]

End LLM
New Query:  {'languageList': ['English'], 'cityList': ['ho chi minh'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['marketing manager', 'content writer', 'marketing assistant', 'marketing analyst', 'financial planner', 'production planner', 'content creator', 'content marketing', 'designer', 'strategic planning', 'marketing officer'], 'role_exp': 1.5, 'skillList': ['social media marketing', 'marketing strategy', 'content creation', 'content writing', 'marketing', 'content marketing', 'digital marketing', 'creative concept development', 'financial planning', 'email marketing', 'value creation', 'decision making', 'content management', 'planning', 'strategic planning', 'production planning', 'building'], 'suitableList': ['insurance advisor', 'marketing executive', 'event coordinator', 'marketing manager', 'content writer', 'marketing assistant', 'marketing analyst', 'content marketing manager', 'financial planner', 'production

 74%|███████▍  | 29/39 [04:13<01:27,  8.77s/it]

End LLM
New Query:  {'institutionList': ['national economics university (neu)'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor']}
Extended Num Application:  30
Not enough application
New Query:  {'institutionList': ['national economics university (neu)', 'university of economics']}
[1214, 1274, 1690, 424, 463, 8, 1941, 2013, 495, 324, 227, 1783, 310, 561, 732, 406, 606, 280, 738, 124, 1761, 1993, 2109, 1722, 1291]
Num CV ok:  25
Rerank
[1214, 124, 1274, 1291, 1690, 1722, 1761, 1783, 1941, 1993, 2013, 2109, 227, 280, 310, 324, 406, 424, 463, 495, 561, 606, 732, 738, 8]
Time: 8.437975645065308
Extracting Job Description


 77%|███████▋  | 30/39 [04:20<01:14,  8.30s/it]

End LLM
New Query:  {'languageList': ['English'], 'institutionList': ['university of economics and law', 'university of economics'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor']}
Extended Num Application:  4
Num CV ok:  4
Rerank
Reranking
[692, 443, 1901]
Time: 7.216526508331299
Extracting Job Description


 79%|███████▉  | 31/39 [04:29<01:09,  8.64s/it]

End LLM
New Query:  {'institutionList': ['hcmc open university', 'ho chi minh city open university'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor']}
Extended Num Application:  11
Not enough application
New Query:  {'institutionList': ['international university - ho chi minh city national university', 'hcmc open university', 'ho chi minh city open university']}
[456, 618, 533, 548, 1669, 1604, 1240, 1193]
Num CV ok:  8
Rerank
[1193, 1240, 1604, 1669, 456, 533, 548, 618]
Time: 9.416627645492554
Extracting Job Description


 82%|████████▏ | 32/39 [04:37<00:58,  8.33s/it]

End LLM
New Query:  {'languageList': ['English'], 'institutionList': ['ho chi minh city university of technology and education', 'ho chi minh city university of technology'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor']}
Extended Num Application:  4
Reduce CV  6  ->  4  CV
New Query:  {'languageList': ['English'], 'institutionList': ['ho chi minh city university of technology and education', 'ho chi minh city university of technology'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor']}
Still more -> Reduce CV
Risk:  True
Num CV ok:  4
Rerank
Reranking
[2027, 2103, 262]
Time: 7.60503625869751
Extracting Job Description


 85%|████████▍ | 33/39 [04:47<00:53,  8.94s/it]

End LLM
New Query:  {'institutionList': ['university of economics and law', 'university of economics'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor']}
Extended Num Application:  3
Reduce CV  4  ->  3  CV
New Query:  {'institutionList': ['university of economics and law'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor']}
Risk:  True
Num CV ok:  3
Rerank
Reranking
[443, 692]
Time: 10.356152057647705
Extracting Job Description


/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/langchain_core/vectorstores.py:330: UserWarning: Relevance scores must be between 0 and 1, got [(Document(page_content='Languages:\n- English\nMajors:\n- bachelor, information security, ho chi minh city university of technology and food industry, (GPA: 2.9)\nWork:\n- công ty c ổ  ph ầ n công ngh ệ tam khoadeveloper .net (1 years)\nRoles:\n- tester (0 years)\n- researcher (0.5 years)\n- software developer (1 years)\nSkills:\n- software development\n- machine learning\n- sql\n- technical support\nProgramming languages:\n- .net (1 years)\n- python (0.5 years)\n- c# (1 years)\n- sql (1 years)\nCities:\n- ho chi minh\n', metadata={'id': 1841}), -196.6845100857687), (Document(page_content='Languages:\n- English\nMajors:\n- bachelor, information technology, vnu-hcm university of science, (GPA: 3.6)\nRoles:\n- software engineer (0.5 years)\n- web developer (0.5 years)\nSkills:\n- cloud storage\n- devops\n- machine learning\n- database 

End LLM
New Query:  {'languageList': ['English'], 'institutionList': ['ho chi minh city university of foreign languages and information technology', 'university of foreign languages - information technology'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor']}
Extended Num Application:  4
Not enough application
New Query:  {'institutionList': ['ho chi minh city university of foreign languages and information technology', 'university of foreign languages - information technology']}
[409, 555]
Less -> Sematic Search
Using Sematic Search
Num CV ok:  3
Rerank
[1841, 409, 555]
Time: 7.437647104263306
Extracting Job Description


 90%|████████▉ | 35/39 [05:04<00:34,  8.64s/it]

End LLM
New Query:  {'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['sales analyst', 'analyst', 'finance analyst', 'business analyst']}
Extended Num Application:  6
Num CV ok:  6
Rerank
Reranking
[548, 443, 1193, 1396]
Time: 9.00951623916626
Extracting Job Description


 92%|█████████▏| 36/39 [05:11<00:24,  8.28s/it]

End LLM
New Query:  {'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor'], 'roleList': ['legal consultant', 'financial consultant', 'investment consultant', 'consultant']}
Extended Num Application:  6
Not enough application
New Query:  {'roleList': ['legal consultant', 'investment consultant', 'consultant', 'financial consultant', 'implementation consultant']}
[669, 548, 1193, 456, 1783]
Num CV ok:  5
Rerank
Reranking
[1193, 1783, 669, 548, 456]
Time: 7.425949573516846
Extracting Job Description


 95%|█████████▍| 37/39 [05:21<00:17,  8.63s/it]

End LLM
New Query:  {'institutionList': ['ho chi minh city university of industry and trade', 'ho chi minh city university of technology and food industry'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor']}
Extended Num Application:  3
Num CV ok:  3
Rerank
Reranking
[1841, 178]
Time: 9.458472728729248
Extracting Job Description


/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/langchain_core/vectorstores.py:330: UserWarning: Relevance scores must be between 0 and 1, got [(Document(page_content='Languages:\n- English\n- Vietnamese\nMajors:\n- undergraduate, computer networking and communication, ho chi minh city university of technology, (GPA: 3.6)\nWork:\n- hung phat usa (0.5 years)\n- private supermarket (0.5 years)\n- mwg (0.5 years)\n- etomil milk (0.5 years)\n- ho chi minh city university of technology (6.5 years)\nRoles:\n- it support (0.5 years)\n- inventory management staff (0.5 years)\n- security camera supervisor (0.5 years)\n- marketing staff (0.5 years)\n- programmer (0.5 years)\nSkills:\n- system configuration\n- management\n- network administration\n- project management\n- web development\n- mobile development\n- game development\n- security\n- iot\nProgramming languages:\n- html (0.5 years)\n- css (0.5 years)\n- unity (0.5 years)\n- wordpress (0.5 years)\n- sql (0.5 years)\nCities:\n- t

End LLM
New Query:  {'institutionList': ['ho chi minh city university of economics'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor']}
Extended Num Application:  11
Not enough application
New Query:  {'institutionList': ['ho chi minh city university of economics', 'ho chi minh city university of science']}
[1253, 1428, 669, 468, 135, 1958]
Less -> Sematic Search
Using Sematic Search
Num CV ok:  8
Rerank
[1253, 135, 1428, 1488, 1530, 1958, 468, 669]
Time: 10.224083423614502
Extracting Job Description


/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/langchain_core/vectorstores.py:330: UserWarning: Relevance scores must be between 0 and 1, got [(Document(page_content='Languages:\n- English\nMajors:\n- bachelor, software engineer, hanoi university of industry, (GPA: 3.5)\nRoles:\n- android developer (0 years)\nSkills:\n- user interface design\n- api design\n- firebase\n- android development\n- git\n- postman\n- mvvm architecture\n- clean architecture\nProgramming languages:\n- kotlin (0 years)\n- dart (0 years)\nCities:\n- ha noi\nAwards:\n- Second Prize in Japanese dubbing contest\n- Consolation Prize in School-level Olympic Programming Contest\n- Third Prize in School-level Olympic Programming Contest\n', metadata={'id': 651}), -160.3873319785448), (Document(page_content='Languages:\n- English\nMajors:\n- bachelor, computer science, national economics university hanoi, (GPA: 3.43)\nWork:\n- lts group (0.5 years)\n- viettel telecom (0.5 years)\nRoles:\n- software developer 

End LLM
New Query:  {'institutionList': ['posts and telecommunications institute of technology', 'vnu-hcm university of science', 'fpt polytechnic hà nội', 'university of technology information - vnu', 'university of information technology'], 'eduList': ['master', 'undergraduate', 'phd', 'diploma', 'certificate', 'bachelor']}
Extended Num Application:  12
Not enough application
New Query:  {'institutionList': ['posts and telecommunications institute of technology', 'vnu-hcm university of science', 'fpt polytechnic hà nội', 'university of technology information - vnu', 'university of information technology']}
[363, 751, 1488, 2006, 2051]
Less -> Sematic Search
Using Sematic Search
Num CV ok:  9
Rerank
[1488, 1530, 1593, 19, 2006, 2051, 363, 651, 751]
Time: 8.397864580154419


In [7]:
recall = 0
precision = 0
long_recall = 0
long_precision = 0
len_long = 0
short_recall = 0
short_precision = 0
len_short = 0

for answer in answers:
    num_ans = len(answer['answer'])
    correct = 0
    ans = set(answer['answer'])
    for p in answer['predict']:
        if p in ans:
            correct += 1
    recall += correct / num_ans
    precision += correct / (len(answer['predict'])+1e-5)
    if num_ans > 7:
        long_recall += correct / num_ans
        long_precision += correct / (len(answer['predict'])+1e-5)
        len_long+= 1
    else:
        short_recall += correct / num_ans
        short_precision += correct / (len(answer['predict'])+1e-5)
        len_short += 1

In [8]:
recall /= len(answers)
precision /= len(answers)
long_recall /= len_long
long_precision /= len_long
short_recall /= len_short
short_precision /= len_short
f1 = 2 * recall * precision / (recall + precision + 1e-5)
short_f1 = 2 * short_recall * short_precision / (short_recall + short_precision + 1e-5)
long_f1 = 2 * long_recall * long_precision / (long_recall + long_precision+1e-5)

In [9]:
print(f"Recall: {recall}")
print(f"Precision: {precision}")
print(f"F1: {f1}")
print("=====================================")

print(f"Short Recall: {short_recall}")
print(f"Short Precision: {short_precision}")
print(f"Short F1: {short_f1}")
print("=====================================")

print(f"Long Recall: {long_recall}")
print(f"Long Precision: {long_precision}")
print(f"Long F1: {long_f1}")
print("=====================================")

Recall: 0.4582659932659932
Precision: 0.45441863442709507
F1: 0.45632920485280953
Short Recall: 0.4209677419354839
Short Precision: 0.42419216192259934
Short F1: 0.4225688012224169
Long Recall: 0.6027967171717171
Long Precision: 0.5715462153820161
Long F1: 0.5867506654834951


## Haiku

Recall: 0.4582659932659932 \
Precision: 0.45441863442709507 \
F1: 0.45632920485280953

=====================================

Short Recall: 0.4209677419354839 \
Short Precision: 0.42419216192259934 \
Short F1: 0.4225688012224169

=====================================

Long Recall: 0.6027967171717171 \
Long Precision: 0.5715462153820161 \
Long F1: 0.5867506654834951

=====================================

## Llama 3 8b

## Llama 3 70b